### DISCUSSION TITLE
Async tasks calling same dbus method simultaneously on Client side

### DISCUSSION THREAD

**Discussion by colazzo** [python-sdbus/python-sdbus#80] [Q&A]
Hi, I have a question regarding async calls.

From my dbus client which uses asyncio:
What happens exactly if I have several asyncio task launched after each other calling the same async method on the dbus proxy from the Client? 
Is there a concurrency check to be done on the dbus Server side, 
or are the received calls being executed one by one on the main thread of the Server?

**Comment by igo95862** [comment]
Hello @colazzo 

> What happens exactly if I have several asyncio task launched after each other calling the same async method on the dbus proxy from the Client?

Each task will create a D-Bus message that will be sent to the D-Bus daemon. D-Bus daemon will when distribute those messages to the address you are calling. I don't think the message delivery order can be reliable.

> Is there a concurrency check to be done on the dbus Server side, or are the received calls being executed one by one on the main thread of the Server?

Is your server also using python-sdbus?

**Reply by colazzo** [reply]
Yes, it also using python-sdbus.
I wondering what really happens with simultaneous calls, do we need to see these the same like having separate async task running at the same time on the servers side? Asking this because then I would need to use asyncio.Locks for critical section...?

**Reply by igo95862** [reply]
The way asyncio works is that it only switches between concurrent tasks when `await` is invoked. If your call handler only uses blocking calls it will never switch to another call mid execution.

It makes sense to use `asyncio.Lock` when you have some shared resource that does not support concurrency.

However, because task switching is predictable and only happens on `await` some read-modify-write loops won't need locking if they are completely blocking.

```python

i = 0

async def test():
    global i
    i = i + 1
```

This function will be concurrency safe in asyncio because it does not use `await` between where it reads the value and modifies it.

**Reply by colazzo** [reply]
@igo95862 Thank you for the clarifications
